rs/node_rewards/canister/src/timer_tasks.rs

```diff
@@ -0,0 +1,133 @@
+use crate::canister::{NodeRewardsCanister, current_time};
+use async_trait::async_trait;
+use ic_cdk_timers::{TimerId, clear_timer};
+use ic_nervous_system_timer_task::{
+    PeriodicSyncTask, RecurringAsyncTask, TimerTaskMetricsRegistry,
+};
+use std::cell::RefCell;
+use std::thread::LocalKey;
+use std::time::Duration;
+
+pub enum TaskStatus {
+    Completed,
+    NotCompleted,
+}
+
+// ================================================================================================
+// DAILY SYNC TASK
+// ================================================================================================
+
+#[derive(Copy, Clone)]
+pub struct DailySyncTask {
+    canister: &'static LocalKey<RefCell<NodeRewardsCanister>>,
+    metrics: &'static LocalKey<RefCell<TimerTaskMetricsRegistry>>,
+}
+
+const DAY_IN_SECONDS: u64 = 60 * 60 * 24;
+const DAILY_SYNC_AT_SECONDS_AFTER_MIDNIGHT: u64 = 5 * 60; // 5 minutes after midnight
+const DAILY_SYNC_RETRY_DELAY: Duration = Duration::from_secs(10 * 60); // 10 minutes
+
+impl DailySyncTask {
+    pub fn new(
+        canister: &'static LocalKey<RefCell<NodeRewardsCanister>>,
+        metrics: &'static LocalKey<RefCell<TimerTaskMetricsRegistry>>,
+    ) -> Self {
+        Self { canister, metrics }
+    }
+
+    fn default_delay(&self) -> Duration {
+        let now_secs = current_time().as_secs_since_unix_epoch();
+        let since_midnight = now_secs % DAY_IN_SECONDS;
+        let mut next_sync_target_secs =
+            now_secs + DAILY_SYNC_AT_SECONDS_AFTER_MIDNIGHT - since_midnight;
+        if since_midnight > DAILY_SYNC_AT_SECONDS_AFTER_MIDNIGHT {
+            next_sync_target_secs += DAY_IN_SECONDS;
+        }
+
+        Duration::from_secs(next_sync_target_secs - now_secs)
+    }
+}
+
+#[async_trait]
+impl RecurringAsyncTask for DailySyncTask {
+    async fn execute(self) -> (Duration, Self) {
+        let registry_sync_result = NodeRewardsCanister::schedule_registry_sync(self.canister).await;
+        let delay = match registry_sync_result {
+            Ok(_) => {
+                ic_cdk::futures::spawn_017_compat(async move {
+                    NodeRewardsCanister::schedule_metrics_sync(self.canister).await
```

Comment:
why are you immediately doing this again?  I don't follow why it needs to happen a second time.